### PR TITLE
Add PVC provisioner for ws projects to openshift infra

### PR DESCRIPTION
**Pull request by akorneta**
Adds provision mechanism for adjusting environment in accordance with infra needs.
Provides persistent volume claims for workspace projects, volume bindings, and volume mounts into Pod and container respectively.
Also added:
- property for disable PVC provisioning:
`CHE_INFRA_OPENSHIFT_WORKSPACE_PVC_ENABLE` with default value `true`
- property for define quantity of PVC:
`CHE_INFRA_OPENSHIFT_WORKSPACE_PVC_QUANTITY` with default value `10Gi`

### CODE REVIEW

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java [hunk 1]
```diff
@@ -200,6 +204,24 @@ private void prepareOpenShiftProject(String projectName) throws InfrastructureEx
...
     }
 
+    private void prepareOpenShiftPVCs(OpenShiftEnvironment osEnv, String namespace) throws InfrastructureException {
+        try (OpenShiftClient client = clientFactory.create()) {
+            final Map<String, PersistentVolumeClaim> existing = new HashMap<>();
+            client.persistentVolumeClaims()
+                  .inNamespace(namespace)
+                  .list()
+                  .getItems()
+                  .forEach(p -> existing.put(p.getMetadata().getName(), p));
```
Comment (benoitf) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java]: Collectors.toMap would have helped ?

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java [hunk 2]
```diff
@@ -200,6 +204,24 @@ private void prepareOpenShiftProject(String projectName) throws InfrastructureEx
...
+            final Map<String, PersistentVolumeClaim> toCreate =
+                    Maps.difference(osEnv.getPersistentVolumeClaims(), existing).entriesOnlyOnLeft();
+            for (PersistentVolumeClaim pvc : toCreate.values()) {
```
Comment (benoitf) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java]: here you're using old (previous JDK8) for each loop, is it wanted .
also, could we filtered directly in the stream ?

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/volume/PersistentVolumeClaimProvisioner.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                          RuntimeIdentity runtimeIdentity) throws InfrastructureException {
+        if (enableWorkspacePVC) {
+            osEnv.getPersistentVolumeClaims()
+                 .put(runtimeIdentity.getWorkspaceId(),
+                      new PersistentVolumeClaimBuilder()
+                              .withNewMetadata()
+                                   .withName(runtimeIdentity.getWorkspaceId())
+                              .endMetadata()
+                              .withNewSpec()
+                                   .withAccessModes("ReadWriteOnce")
```
Comment (benoitf) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/volume/PersistentVolumeClaimProvisioner.java]: it' not configurable/property ?

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/InfrastructureProvisioner.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+     * @param environment
+     *         configuration of environment
+     * @param osEnv
+     *         OpenShiftEnvironment
+     * @param identity
+     *         runtime identity
+     * @throws InfrastructureException
+     *         when any error occurs
+     */
+    void provision(Environment environment,
```
Comment (garagatyi) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/InfrastructureProvisioner.java]: I think EnvironmentImpl should be used instead since it can be changed according to javadocs

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/codenvy/che.properties [hunk 5]
```diff
@@ -401,3 +401,5 @@ che.infra.openshift.bootstrapper.binary_url=http://che-host:${CHE_PORT}/agent-bi
 che.infra.openshift.bootstrapper.timeout_min=10
 che.infra.openshift.bootstrapper.installer_timeout_sec=180
 che.infra.openshift.bootstrapper.server_check_period_sec=3
+che.infra.openshift.workspace.pvc.enable=true
```
Comment (garagatyi) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/codenvy/che.properties]: I think word 'workspace' can be removed from these properties

**Review by benoitf — APPROVED** [review]

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInfraModule.java [hunk 6]
```diff
@@ -28,6 +30,9 @@ protected void configure() {
                                                                                       RuntimeInfrastructure.class);
         infrastructures.addBinding().to(OpenShiftInfrastructure.class);
 
+        bind(InfrastructureProvisioner.class).to(OpenShiftInfrastructureProvisioner.class);
+        bind(ConfigurationProvisioner.class).to(PersistentVolumeClaimProvisioner.class);
```
Comment (sleshchenko) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInfraModule.java]: I guess this binding is useless

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java [hunk 7]
```diff
@@ -277,10 +301,10 @@ private void sendRunningEvent(String machineName) {
                                        .withMachineName(machineName));
     }
 
-    private static class DeleteWatcher implements Watcher<Pod> {
+    public static class DeleteWatcher implements Watcher<Pod> {
```
Comment (sleshchenko) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java]: Why did you change it?

**Review by riuvshin — APPROVED** [review]

**Review by skabashnyuk — APPROVED** [review]

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/codenvy/che.properties [hunk 8]
```diff
@@ -401,3 +401,6 @@ che.infra.openshift.bootstrapper.binary_url=http://che-host:${CHE_PORT}/agent-bi
 che.infra.openshift.bootstrapper.timeout_min=10
 che.infra.openshift.bootstrapper.installer_timeout_sec=180
 che.infra.openshift.bootstrapper.server_check_period_sec=3
+che.infra.openshift.pvc.enable=true
```
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/codenvy/che.properties]: che.infra.openshift.pvc.enable**d**

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/codenvy/che.properties [hunk 9]
```diff
@@ -401,3 +401,6 @@ che.infra.openshift.bootstrapper.binary_url=http://che-host:${CHE_PORT}/agent-bi
...
+che.infra.openshift.pvc.quantity=10Gi
```
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/codenvy/che.properties]: I'd add comment about format for value of this property

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInfraModule.java [hunk 10]
```diff
@@ -28,6 +28,8 @@ protected void configure() {
                                                                                       RuntimeInfrastructure.class);
         infrastructures.addBinding().to(OpenShiftInfrastructure.class);
 
+        bind(OpenShiftInfrastructureProvisioner.class);
```
Comment (sleshchenko) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInfraModule.java]: I'd say that this binding is useless

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/volume/PersistentVolumeClaimProvisioner.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    @Override
+    public void provision(EnvironmentImpl environment,
+                          OpenShiftEnvironment osEnv,
+                          RuntimeIdentity runtimeIdentity) throws InfrastructureException {
+        if (pvcEnable) {
+            osEnv.getPersistentVolumeClaims()
+                 .put(runtimeIdentity.getWorkspaceId(),
+                      new PersistentVolumeClaimBuilder()
+                              .withNewMetadata()
+                                   .withName(runtimeIdentity.getWorkspaceId())
```
Comment (sleshchenko) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/volume/PersistentVolumeClaimProvisioner.java]: I'd use something like `wsId-projects` for name of this pvc. WDYT?

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInfrastructureProvisioner.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.openshift.environment.OpenShiftEnvironment;
+import org.eclipse.che.workspace.infrastructure.openshift.provision.installer.InstallerConfigProvisioner;
+import org.eclipse.che.workspace.infrastructure.openshift.provision.volume.PersistentVolumeClaimProvisioner;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * @author Anton Korneta
```
Comment (garagatyi) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInfrastructureProvisioner.java]: Please, add javadocs

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInfraModule.java (hunk 10 above)
Comment (garagatyi) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInfraModule.java]: You don't have to bind this class explicitly

File: infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInfrastructureProvisionerTest.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        osInfraProvisioner.provision(environment, osEnv, runtimeIdentity);
+
+        provisionOrder.verify(installerProvisioner)
+                      .provision(eq(environment), eq(osEnv), eq(runtimeIdentity));
+        provisionOrder.verify(pvcProvisioner)
+                      .provision(eq(environment), eq(osEnv), eq(runtimeIdentity));
+        provisionOrder.verifyNoMoreInteractions();
+    }
+
+}
```
Comment (garagatyi) [inline infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInfrastructureProvisionerTest.java]: Please add trailing new line at the end of file

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java [hunk 14]
```diff
@@ -200,6 +203,27 @@ private void prepareOpenShiftProject(String projectName) throws InfrastructureEx
         }
     }
 
+    private void prepareOpenShiftPVCs(OpenShiftEnvironment osEnv, String namespace) throws InfrastructureException {
```
Comment (garagatyi) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java:206]: Looks like this part is not covered by tests. Are you going to add tests in this PR?

Comment (akorneta) [reply infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java:206]: In this pr that part won't be covered

**Review by garagatyi — APPROVED** [review]

File: assembly/assembly-main/pom.xml [hunk 15]
```diff
@@ -65,11 +65,11 @@
             <artifactId>che-core-ide-templates</artifactId>
             <type>jar</type>
         </dependency>
-        <dependency>
-            <groupId>org.eclipse.che.dashboard</groupId>
-            <artifactId>che-dashboard-war</artifactId>
-            <type>war</type>
-        </dependency>
+        <!--<dependency>-->
```
Comment (sleshchenko) [inline assembly/assembly-main/pom.xml]: Don't forget to uncomment it before merge

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/codenvy/che.properties [hunk 16]
```diff
@@ -401,3 +401,17 @@ che.infra.openshift.bootstrapper.binary_url=http://che-host:${CHE_PORT}/agent-bi
...
+
+# Defines whether use the Persistent Volume Claim for che workspace needs
+# e.g backup projects, logs etc or disable it.
+che.infra.openshift.pvc.enabled=true
+# Defines the name of Persistent Volume Claim for che workspace.
+che.infra.openshift.pvc.name=claim-che-workspace
+# Defines the size of Persistent Volume Claim of che workspace.
+# Format described here:
+# https://docs.openshift.com/container-platform/latest/dev_guide/compute_resources.html#dev-compute-resources
+che.infra.openshift.pvc.quantity=10Gi
```
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/codenvy/che.properties:413]: Mb it make sense to rename it to something like `che.infra.openshift.projects.pvc.name` while this pvc is for projects sources only

**Review by sleshchenko — APPROVED** [review]
OK for me
